src/models/contact/index.js

```diff
@@ -71,19 +65,27 @@ export async function getContactAll(addressBook, types, fetch) {
 }
 
 export async function addContactIndexToAddressBook(addressBook, type, fetch) {
-  const indexUrl = getContactIndexDefaultUrl(addressBook.containerUrl, type);
-  const datasetUrl = getAddressBookIndexUrl(addressBook);
-  const dataset = await saveSolidDatasetAt(
-    datasetUrl,
-    setThing(
-      addressBook.dataset,
-      setUrl(addressBook.thing, type.indexFilePredicate, indexUrl)
-    ),
-    { fetch }
-  );
+  let updatedAddressBook = addressBook;
+  let indexUrl = getContactIndexUrl(addressBook, type);
+  if (!indexUrl) {
+    const newIndexUrl = getContactIndexDefaultUrl(
+      addressBook.containerUrl,
+      type
+    );
+    const datasetUrl = getAddressBookIndexUrl(addressBook);
+    updatedAddressBook = await saveSolidDatasetAt(
```

Comment:
Oh, sorry, didn't see this earlier. You're changing the type of `updatedAddressBook` here, from the model `AddressBook` to `SolidDataset`. This won't work.

Make sure to package a new object `{ containerUrl, dataset, thing }` that uses the dataset returned from `saveSolidDatasetAt`:

```
export async function addContactIndexToAddressBook(addressBook, type, fetch) {
  let updatedAddressBook = addressBook;
  let indexUrl = getContactIndexUrl(addressBook, type);
  if (!indexUrl) {
    const newIndexUrl = getContactIndexDefaultUrl(
      addressBook.containerUrl,
      type
    );
    const datasetUrl = getAddressBookIndexUrl(addressBook);
    const updatedAddressBookDataset = await saveSolidDatasetAt(
      datasetUrl,
      setThing(
        addressBook.dataset,
        setUrl(addressBook.thing, type.indexFilePredicate, newIndexUrl)
      ),
      { fetch }
    );
    updatedAddressBook = {
      containerUrl: addressBook.containerUrl,
      dataset: updatedAddressBookDataset,
      thing: getThing(updatedAddressBookDataset, asUrl(addressBook.thing))
    };
    indexUrl = getContactIndexUrl(updatedAddressBook, type);
  }

  return {
    addressBook: updatedAddressBook,
    indexUrl,
  };
}
```